File: 04_Stack/04_Generate_Parentheses/0022-generate-parentheses.cpp
/*
Problem: LeetCode 22 - Generate Parentheses

Description:
Given an integer n, generate all combinations of well-formed parentheses of length 2n.

Intuition:
To generate all combinations of well-formed parentheses, we can use a backtracking approach. At each step, we have two choices: either to place an opening parenthesis or a closing parenthesis.

Approach:
1. Initialize an empty result vector to store all valid combinations.
2. Define a helper function, backtrack, that takes the current combination, the count of opening parentheses, the count of closing parentheses, and the result vector.
3. In the backtrack function:
   - If the length of the current combination is equal to 2n, add it to the result vector.
   - If the count of opening parentheses is less than n, recursively call the backtrack function with the current combination appended with an opening parenthesis and an incremented count of opening parentheses.
   - If the count of closing parentheses is less than the count of opening parentheses, recursively call the backtrack function with the current combination appended with a closing parenthesis and an incremented count of closing parentheses.
4. Call the backtrack function initially with an empty combination, 0 opening parentheses, 0 closing parentheses, and the result vector.
5. Return the result vector containing all valid combinations.

Time Complexity:
The time complexity is O(4^n / sqrt(n)), as there are Catalan numbers of well-formed parentheses combinations.

Space Complexity:
The space complexity is O(4^n / sqrt(n)), as there can be a total of 4^n / sqrt(n) combinations generated.
*/

class Solution {
  public:
    vector<string> generateParenthesis(int n) {
        vector<string> result;
        generate(n, 0, 0, "", result);
        return result;
    }
  private:
    void generate(int n, int open, int close, string str, vector<string> &result) {
        if (open == n && close == n) {
            result.push_back(str);
            return;
        }

        if (open < n) {
            generate(n, open + 1, close, str + '(', result);
        }

        if (open > close) {
            generate(n, open, close + 1, str + ')', result);
        }
    }
};

File: 04_Stack/04_Generate_Parentheses/0022-generate-parentheses.rs
/*
Problem: LeetCode 22 - Generate Parentheses

Key Idea:
The key idea is to use recursion to generate all valid combinations of parentheses.

Approach:
1. Initialize an empty vector `result` to store the generated combinations.
2. Create a recursive function `generate` that takes four arguments: `current` (the current combination), `open` (the number of open parentheses used), `close` (the number of closing parentheses used), and `n` (the total number of pairs).
3. In the `generate` function:
   - If `current` is of length `2 * n`, it means a valid combination is generated, so add it to the `result` vector.
   - If `open` is less than `n`, it means we can add an open parenthesis. Recursively call `generate` with `current` + '(', `open + 1`, `close`, and `n`.
   - If `close` is less than `open`, it means we can add a closing parenthesis. Recursively call `generate` with `current` + ')', `open`, `close + 1`, and `n`.
4. Start the recursion with an empty `current`, `open` and `close` as 0, and `n` as the given input.
5. Return the `result` vector containing all valid combinations.

Time Complexity:
O(2^n), where n is the number of pairs. Each valid combination has a length of 2 * n, and there are 2^(2n) possible combinations.

Space Complexity:
O(2^n), as there can be a total of 2^(2n) valid combinations in the result vector.
*/

impl Solution {
    pub fn generate_parenthesis(n: i32) -> Vec<String> {
        let mut result: Vec<String> = Vec::new();

        fn generate(current: String, open: i32, close: i32, n: i32, result: &mut Vec<String>) {
            if current.len() == (2 * n) as usize {
                result.push(current);
                return;
            }

            if open < n {
                generate(current.clone() + "(", open + 1, close, n, result);
            }

            if close < open {
                generate(current.clone() + ")", open, close + 1, n, result);
            }
        }

        generate("".to_string(), 0, 0, n, &mut result);
        result
    }
}
